### ISSUE TITLE
Experiment with EXCHANGE operator for JOIN and GROUP BY

### ISSUE THREAD

**Issue by kitaisreal** [ClickHouse/ClickHouse#53451]
EXCHANGE operator with hash partitioning scheme can be used to increase parallelism for JOIN and GROUP BY operators in cases of high data cardinality.

## GROUP BY

Currently for aggregation with high data cardinality we use 2 level hash table and parallelize each step of aggregation. N threads build 2 level hash table with 256 buckets during first stage of aggregation. On the next stage multiple buckets with the same bucket number from different threads are merged into single bucket for each bucket number in parallel. And after that buckets are finalized in parallel.
When there is high cardinality in data with such approach size of each hash table in each thread can become huge and merge stage can take significant amount of time. 
The idea is to first PARTITION data by hash, so that each thread will work with hash table of less size and to avoid merge stage completely.

## JOIN

For JOIN hash partitioning can improve several scenarious:

1. It can be used as building block for `ConcurrentHashJoin` and `GraceHashJoin` and potentially improve performance because partitioning will be implemented on top of query pipeline.
2. For `FullSortingMergeJoin` we need to sort both tables before joining them, or when both tables are sorted we only need to merge data. The final merge stage is single threaded. We can improve that using HASH parititoning before sorting or merging data, that way merge stage will be parallelized across all threads.

**Comment by UnamedRus** [comment]
It can be also integrated with aggregation_in_order reading? 
As column values in ORDER BY are naturally can be split in non intersecting ranges?

https://github.com/ClickHouse/ClickHouse/pull/36396#issuecomment-1112146357

**Comment by ucasfl** [comment]
It the table is a Distributed table, how the data will be partitioned? And can it be extend to support distributed shuffle join?

**Comment by alexey-milovidov** [comment]
Duplicate of https://github.com/ClickHouse/ClickHouse/issues/74123.
